closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java

```diff
@@ -460,6 +508,65 @@ int decryptFinal(byte[] in, int inOfs, int len,
         return ret;
     }
 
+
+    /**
+     * Performs decryption operation for the last time.
+     *
+     * @param src the input buffer with the data to be decrypted
+     * @param dst the offset in <code>in</code>
+     * @return the number of bytes placed into the <code>out</code> buffer
+     */
+    int decryptFinal(ByteBuffer src, ByteBuffer dst)
+        throws IllegalBlockSizeException, AEADBadTagException,
+        ShortBufferException {
+
+        // Get array from source
+        byte[] src_arr = getBbArray(src);
+        int src_ofs = src.position() + (src.hasArray() ? src.arrayOffset() : 0);
+
+        // Get array from destination
+        byte[] dst_arr;
+        if (dst.hasArray()) {
+            dst_arr = dst.array();
+        } else {
+            dst_arr = new byte[dst.capacity()];
+        }
+        int dst_offset = dst.position() + (dst.hasArray() ? dst.arrayOffset() : 0);
+
+        int len = decryptFinal(src_arr, src_ofs, src.remaining(), dst_arr, dst_offset);
+        
+        // Update destination buffer
+        if (dst.hasArray()) {
+            dst.position(dst.position() + len);
+            dst.limit(dst.position());
+        } else {
+            dst.put(dst_arr, dst.position(), len);       
+        }
+        
+        return len;
+    }
+
+    /**
+     * Gets the byte array behind a buffer.
+     * Tries to use ByteBuffer.array(). If this is not available, the function uses ByteBuffer.get()
+     *
+     * @param src the buffer whose byte[] is needed
+     * @return the byte array with the buffer's content
+     */
+    private byte[] getBbArray(ByteBuffer src) {
+        byte[] arr;
+        if (src.hasArray()) {
+            arr = src.array();
+        } else {
+            ByteBuffer cpy = src.duplicate();
+            arr = new byte[cpy.capacity()];
+            cpy.limit(cpy.capacity());
+            cpy.position(0);
+            if (arr.length > 0) cpy.get(arr, 0, cpy.capacity());
```

Comment:
I don't think this is correct. The length of the resulting array should be `src.remaining()`.
```
            ByteBuffer cpy = src.duplicate();
            arr = new byte[cpy.remaining()];
            if (arr.length > 0) {
                cpy.get(arr, 0, arr.length);
            }
```
This helper method can also be `static`.

Reply:
The purpose of this function is to get an array similar to the one obtained from `buffer.array()` when the buffer is array-backed. Then I control the memory used in this array using the offset and length. The reason I do this is to get the same array structure regardless of whether the buffer is array-backed or not

Reply:
I thought about this more: I don't think it's reasonable to create a byte array of length `buffer.remaining()`; imagine the case where the `ByteBuffer` is mapped to a large input file.